analysis/skeleton_length_analysis.py

```diff
@@ -0,0 +1,217 @@
+                error = abs(len(plan_skeleton) - demo_len)
+                task_skeleton_length_errors.append(error)
+        except (ApproachTimeout, ApproachFailure):
+            # Use an upper bound on the error.
+            num_missing = max_skeletons - len(task_skeleton_length_errors)
+            for _ in range(num_missing):
+                task_skeleton_length_errors.append(error_upper_bound)
+        skeleton_length_errors.append(task_skeleton_length_errors)
+    errors_arr = np.array(skeleton_length_errors, dtype=np.int32)
+    assert errors_arr.shape == (len(train_tasks), max_skeletons)
+    return errors_arr
+
+
+def _create_heatmap(env_results: NDArray[np.int32], env_name: str,
+                    outfile: str) -> None:
+    # Env results shape is (seed, predicate set, task, skeleton idx).
+    # Reorganize into heatmap array of shape (predicate set, skeleton idx)
+    # by averaging out seed and task.
+    heatmap_arr = np.mean(env_results, axis=(0, 2))
+    num_predicate_sets, num_skeletons = heatmap_arr.shape
+
+    fig, ax = plt.subplots()
+    ax.imshow(heatmap_arr.T, origin='lower')
+    ax.set_xticks(np.arange(num_predicate_sets))
+    ax.set_yticks(np.arange(num_skeletons))
+    for i in range(num_skeletons):
+        for j in range(num_predicate_sets):
+            ax.text(j,
+                    i,
+                    heatmap_arr[j, i],
+                    ha="center",
+                    va="center",
+                    color="w")
+
+    ax.set_title(f"{env_name.capitalize()}: Skeleton Length Errors")
+    ax.set_xlabel("Num Non-Goal Predicates Included")
+    ax.set_ylabel("Skeleton Index")
+    fig.tight_layout()
+    plt.savefig(outfile)
+    print(f"Wrote out to {outfile}.")
+
+
+def _create_plot(env_results: NDArray[np.int32], env_name: str,
+                 outfile: str) -> None:
+    # Env results shape is (seed, predicate set, task, skeleton idx).
+    # Reorganize into array of shape (predicate set,) by minning over skeleton
+    # idx and averaging out seed and task.
+    arr = np.mean(np.min(env_results, axis=3), axis=(0, 2))  # type: ignore
+    num_predicate_sets, = arr.shape
+
+    fig, ax = plt.subplots()
+    ax.plot(np.arange(num_predicate_sets), arr)
+
+    ax.set_xticks(np.arange(num_predicate_sets))
+    ax.set_title(f"{env_name.capitalize()}: Min Skeleton Length Errors")
+    ax.set_xlabel("Num Non-Goal Predicates Included")
+    ax.set_ylabel("Min Skeleton Length Error")
+    fig.tight_layout()
+    plt.savefig(outfile)
+    print(f"Wrote out to {outfile}.")
+
+
+def _main() -> None:
+    outdir = os.path.join(os.path.dirname(os.path.realpath(__file__)),
+                          "results")
+    os.makedirs(outdir, exist_ok=True)
+
+    for env_name in ENV_NAMES:
+        env_results = []
+        for seed in SEEDS:
+            seed_results = _run_analysis_for_env(env_name, seed)
+            env_results.append(seed_results)
+        env_results_arr = np.array(env_results, dtype=np.int32)
```

Comment:
glad there's no weirdness when np.mean()'ing an int array... i would have thought it truncates & makes a new int array...